virtcontainers/qemu.go

```diff
@@ -1191,6 +1229,76 @@ func (q *qemu) disconnect() {
 	q.qmpShutdown()
 }
 
+// resizeMemory get a request to update the VM memory to reqMemMB
+// Memory update is managed with two approaches
+// Add memory to VM:
+// When memory is required to be added we hotplug memory
+// Remove Memory from VM/ Return memory to host.
+//
+// Memory unplug can be slow and it cannot be guaranteed.
+// Additionally, the unplug has not small granularly it has to be
+// the memory to remove has to be at least the size of one slot.
+// To return memory back we are resizing the VM memory balloon.
+// A longer term solution is evaluate solutions like virtio-mem
+func (q *qemu) resizeMemory(reqMemMB uint32, memoryBlockSizeMB uint32) (uint32, error) {
+
+	currentMemory := q.config.MemorySize + uint32(q.state.HotpluggedMemory)
+	err := q.qmpSetup()
+	if err != nil {
+		return 0, err
+	}
+	switch {
+	case currentMemory < reqMemMB:
+		//hotplug
+		addMemMB := reqMemMB - currentMemory
+		memHotplugMB, err := memHotplugValid(addMemMB, memoryBlockSizeMB)
+		if err != nil {
+			return currentMemory, err
+		}
+
+		addMemDevice := &memoryDevice{
+			sizeMB: int(memHotplugMB),
+		}
+		data, err := q.hotplugAddDevice(addMemDevice, memoryDev)
+		if err != nil {
+			return currentMemory, err
+		}
+		memoryAdded, ok := data.(int)
+		if !ok {
+			return currentMemory, fmt.Errorf("Could not get the memory added, got %+v", data)
+		}
+		currentMemory += uint32(memoryAdded)
+	case currentMemory > reqMemMB:
+		//hotunplug
+		addMemMB := currentMemory - reqMemMB
+		memHotunplugMB, err := memHotplugValid(addMemMB, memoryBlockSizeMB)
+		if err != nil {
+			return currentMemory, err
+		}
+
+		addMemDevice := &memoryDevice{
+			sizeMB: int(memHotunplugMB),
+		}
+		data, err := q.hotplugRemoveDevice(addMemDevice, memoryDev)
+		if err != nil {
+			return currentMemory, err
+		}
+		memoryRemoved, ok := data.(int)
+		if !ok {
+			return currentMemory, fmt.Errorf("Could not get the memory removed, got %+v", data)
+		}
+		currentMemory -= uint32(memoryRemoved)
+	}
+
+	// Update balloon in any memory update case with the size requested by the caller.
+	if err := q.updateMemoryBalloon(reqMemMB); err != nil {
```

Comment:
I guess you forgot to add `case currentMemory == reqMemMB` , right?

Reply:
This include all the cases. The hotplug or hot-remove is handled in the `switch` statement and as a separated case the ballloon is always updated with the size of the memory requested by the caller..

Reply:
If `currentMemory == reqMemMB`, `resizeMemory` should be noop. Please add a check at the beginning of the function to make sure we do nothing in such case (which means we do not even setup the qmp connection).